### PR TITLE
Integration testing

### PR DESCRIPTION
**Pull request by TheRealFalcon**
Follow up to #488 . For background and context, see that PR.

Since then, I have filled out the implementation and re-designed a few things based on feedback. Specifically:

- Transitioned to relying on marks instead of a context manager for cloud setup
- Added file push and pull support
- Added support for supplying a cloud config (userdata) via mark
- Added implementations for various image generation types
- Added a few more clouds (Azure still not quite ready)

I still need documentation :slightly_smiling_face:

### CODE REVIEW

File: tests/integration_tests/platforms.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        remote_script = (
+            '{sudo} echo deb "http://archive.ubuntu.com/ubuntu '
+            '$(lsb_release -sc)-proposed main" | '
+            '{sudo} tee /etc/apt/sources.list.d/proposed.list\n'
+            '{sudo} apt-get update -q\n'
+            '{sudo} apt-get install -qy cloud-init'
+        ).format(sudo='sudo' if self.use_sudo else '')
+        self._install_new_cloud_init(remote_script)
+
+    def install_ppa(self):
```
Comment (blackboxsw) [inline tests/integration_tests/platforms.py:126]: Should this inspect an environment variable or cmdline param for which PPA to use. Currently it sets up proposed instead of installing a ppa.

Comment (TheRealFalcon) [reply tests/integration_tests/platforms.py:126]: Isn't this how we install the dev-proposed PPA?

I was thinking mostly along the lines of when we SRU. Before it gets to proposed, we usually push a package to the dev-proposed PPA, right? How do we install that PPA?

Comment (blackboxsw) [reply tests/integration_tests/platforms.py:126]: I was thinking more along the lines or making the PPA option a parameter that we can fill in explicitly. It'd give us more options if we want to test our own PPA for instance.
See my updated comment above.

File: tests/integration_tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        if not isinstance(client, LxdContainerClient):
+            raise ValueError('IN_PLACE as IMAGE_SOURCE only works for LXD')
+        # The mount needs to happen after the instance is launched, so
+        # no further action needed here
+    elif integration_settings.IMAGE_SOURCE == 'PROPOSED':
+        client.launch()
+        client.install_proposed_image()
+    elif integration_settings.IMAGE_SOURCE == 'PPA':
+        client.launch()
+        client.install_ppa()
```
Comment (blackboxsw) [inline tests/integration_tests/conftest.py:56]: I was thinking we could make IMAGE_SOURCE much more flexible here and pass that PPA value into install_ppa()
We could then opt to set IMAGES_SOURCE=ppa:cloud-init-dev/proposed 

```suggestion
    elif integration_settings.IMAGE_SOURCE.startswith("ppa:"):
        client.launch()
        client.install_ppa(ntegration_settings.IMAGE_SOURCE)
```

File: tests/integration_tests/integration_settings.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+# USER SETTINGS OVERRIDES
+##################################################################
+# Bring in any user-file defined settings
+try:
+    from tests.integration_tests.user_settings import *
+except:  # noqa
+    pass
+
+##################################################################
+# ENVIRONMENT SETTINGS OVERRIDES
```
Comment (blackboxsw) [inline tests/integration_tests/integration_settings.py:85]: At some point in might be worth providing a basic argparse for these switches too (as setting env variables CLOUD_INIT_* for any param is bound to get verbose fairly quickly.

File: tests/integration_tests/test_simple.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+CLOUD_CONFIG = """#cloud-config
+bootcmd:
+  - echo 'hello config!' > /tmp/cloud_config.txt"""
+
+
+class TestSimple:
+    @pytest.mark.cloud_config(CLOUD_CONFIG)
+    def test_cloud_config(self, client):
+        client.push_to_file('/home/ubuntu/my_test.txt', 'Hello world!')
+        print(client.exec('cloud-init -v'))
```
Comment (blackboxsw) [inline tests/integration_tests/test_simple.py:19]: Since these are tests, let's assert the heck outta this stuff instead of prints. We can leave the introductory prints in for documentation. But best to show a suggestion of how we expect these tests to look (a bit).

File: tests/integration_tests/test_simple.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+
+CLOUD_CONFIG = """#cloud-config
+bootcmd:
+  - echo 'hello config!' > /tmp/cloud_config.txt"""
+
+
+class TestSimple:
+    @pytest.mark.cloud_config(CLOUD_CONFIG)
+    def test_cloud_config(self, client):
```
Comment (blackboxsw) [inline tests/integration_tests/test_simple.py:17]: If we have multiple tests that we want to run against a single instance, will this current design allow for that?

**Review by blackboxsw — DISMISSED** [review]
Looks pretty good for representing individual tests which could be limited to specific platforms. 

I think if we are looking at gaps vs the current cloud_tests:
  - If we want to perform multiple, separate tests against a single running instance how would that look?
      1. I'm thinking about a case where I share a single instance at the class level scope and run X number of tests against it instead of having to build fairly large single test method which asserts multiple things.
  - I'm wondering  how you would envision support for redhat or sles:
       1. do you imagine that  IntegrationClient would grow more general distro-aware/specialized 'install_package' instead of install_deb
       2. Does the use of pycloudlib also imply that we would need more RHEL/SLES/DEBIAN/FreeBSD support in pycloudlib to support integration tests for cloud-init?

File: tests/integration_tests/test_simple.py (hunk 4 above)
Comment (TheRealFalcon) [reply tests/integration_tests/test_simple.py:19]: Yeah, this is just an example file for playing around with concepts. I don't plan on merging it. I'll come up with some other examples too.

File: tests/integration_tests/test_simple.py (hunk 5 above)
Comment (TheRealFalcon) [reply tests/integration_tests/test_simple.py:17]: Yeah, it's requires some small shenanigans in our conftest. I'll add them.

File: tests/integration_tests/integration_settings.py (hunk 3 above)
Comment (TheRealFalcon) [reply tests/integration_tests/integration_settings.py:85]: I'd actually rather not for a few reasons. 

For customizing things on your individual machine, you can use the file I added to the gitignore: https://github.com/canonical/cloud-init/pull/581#diff-a084b794bc0759e7a6b77810e01874f2R32
Anything in integration_settings.py will be overridden by anything in user_settings.py. Anything in either of those two  will be overridden by anything in an env var.

If we're talking about scripting or running on CI, I think env vars are easier. You can specify them on different lines or source an entire file of them, rather than relying on a single line of a million variable substitutions.

Additionally, since this is being run via pytest, you can't really use argparse directly. You have to hook into pytest. The `pytest -h` output is already fairly verbose, and adding more flags won't exactly make it more user friendly. Also, I wouldn't want them to be displayed when we're running our unit tests, and I'm not sure how easy it would be to customize that.

**Comment by TheRealFalcon** [comment]
> If we want to perform multiple, separate tests against a single running instance how would that look?
        I'm thinking about a case where I share a single instance at the class level scope and run X number of tests against it instead of having to build fairly large single test method which asserts multiple things.

I updated the conftest to add that. It's not the most elegant (client, class_client, module_client), but that's the recommended way of doing it from pytest.

> I'm wondering how you would envision support for redhat or sles:
        do you imagine that IntegrationClient would grow more general distro-aware/specialized 'install_package' instead of install_deb

I haven't really considered that and it's a conversation we need to have. I was mainly approaching this from an Ubuntu SRU mindset. I'm assuming other distros have their own ways of doing integration testing. Do we have any insight into that? Would they want to migrate to what we're doing here? I wouldn't want to attach every distro command to IntegrationClient. I would want to separate that out somehow, but I don't think that needs to be figured out right now. 

> Does the use of pycloudlib also imply that we would need more RHEL/SLES/DEBIAN/FreeBSD support in pycloudlib to support integration tests for cloud-init?

If we're going to support these other distros, then yes, I would think it would make the most sense to push that functionality into pycloudlib (or re-evaluate libcloud :wink: ).

I'm hoping that we can write tests now, and figure out the answers to some of these questions as we go. The tests themselves don't need to be worried about how an instance was launched or what image it's using, unless the test is specific to that configuration in which case we have pytest marks to signify that. If we support other clouds or other distros, the tests themselves shouldn't need change.

File: tests/integration_tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        # one around as it was just for image creation
+        client.destroy()
+    print('Done with environment setup')
+
+
+@contextmanager
+def _client(request, fixture_utils):
+    cloud_config = fixture_utils.closest_marker_first_arg_or(
+        request, 'cloud_config', None)
+    with dynamic_client(user_data=cloud_config) as instance:
```
Comment (OddBloke) [inline tests/integration_tests/conftest.py:72]: We should probably consistently refer to this as either `user_data` or `cloud_config` in both test code and the test infrastructure code.

Comment (TheRealFalcon) [reply tests/integration_tests/conftest.py:72]: @OddBloke Have a preference?

Comment (blackboxsw) [reply tests/integration_tests/conftest.py:72]: I'd vote user_data as I'd like to be able to provide more than just cloud-config to the instance (sample boot scripts #includes etc) which is broader than cloud_config

**Review by lucasmoura — COMMENTED** [review]
That seems really good @TheRealFalcon :clap: :clap: 

My only concern at the moment is regarding the cleanup part of the test. I understand that we are deleting the instance at the end of each test (depending on the client fixture scope), but I am afraid that this may not be enough. For example, in `Azure`,  every time we launch an image, we create a `resource group`  for it. Since we are creating a new client every time we launch an image, we will end up having a lot of `resource groups` not being cleaned up. I don't know if that is the case for `EC2` and `GCE`  too.

What I think we can do is let each cloud client handle the cleanup process, like Azure could call `delete_resource_group` on the `__exit__`  method if we ask to no keep the instance.

Another approach is to reuse the cloud clients. For Azure, we will always be using the same `resource_group` if we use the same client to launch instances. For `EC2`, it seems that it already uses a default `vpc` to create the instances on, so that may not be an issue. But I don't know about the other clouds.

Sorry, for only providing Azure examples here, but if the other clouds do not have a similar behavior, we can only modify Azure on that regard.

**Review by OddBloke — COMMENTED** [review]
Thanks for all your work on this, James, this is looking really good! I have a bunch of thoughts, comments and suggestions inline.

File: integration-requirements.txt [hunk 7]
```diff
@@ -4,6 +4,8 @@
 # Note: Changes to this requirements may require updates to
 # the packages/pkg-deps.json file as well.
 #
+pytest
+git+https://github.com/canonical/pycloudlib.git
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply integration-requirements.txt:8]: I agree, but we need to start releasing new pycloudlib versions in order to have any sort of meaningful pinning. The last release was in 2018.

File: tests/integration_tests/conftest.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        client.install_proposed_image()
+    elif integration_settings.IMAGE_SOURCE.startswith('ppa:'):
+        client.launch()
+        client.install_ppa(integration_settings.IMAGE_SOURCE)
+    elif os.path.isfile(str(integration_settings.IMAGE_SOURCE)):
+        client.launch()
+        client.install_deb()
+    if client.instance:
+        # Even if we're keeping instances, we don't want to keep this
+        # one around as it was just for image creation
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/conftest.py:63]: You'd comment out the terminate line in the image creation fixture :wink: 

I think this is actually reasonable though, because if you're to the point where you're debugging the image snapshot, that means you're inevitably going to be in the fixture/IntegrationClient code.

File: tests/integration_tests/integration_settings.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+##################################################################
+# GCE SPECIFIC SETTINGS
+##################################################################
+# Required for GCE
+GCE_PROJECT = None
+
+# You probably want to override these
+GCE_REGION = 'us-central1'
+GCE_ZONE = 'a'
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/integration_settings.py:65]: `pycloudlib` doesn't do anything with defaulting. You need to pass the values to the library itself through the various constructors. I think we could allow specifying a configuration file to find the setting in. I wouldn't want to automatically use a pre-configured default though. People might have different accounts for different projects, or accidentally typed `pytest` in the root directory without understanding the implications.

File: tests/integration_tests/platforms.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+try:
+    from typing import Callable, Optional
+except ImportError:
+    pass
+
+current_image = None
+
+
+def set_current_image(image):
+    global current_image
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/platforms.py:23]: It'd have to be a class attribute, but sure

File: tests/integration_tests/conftest.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+
+@pytest.fixture(scope='session', autouse=True)
+def setup_image():
+    """Setup the target environment with the correct version of cloud-init.
+
+    So we can launch instances / run tests with the correct image
+    """
+    client = dynamic_client()
+    print('Setting up environment for {}'.format(client.datasource))
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/conftest.py:44]: When it comes to tests (and pytest in particular), I actually prefer printing. I have yet to see a use-case for multiple log levels in tests. I either want everything, or I want nothing. The `-s` flag gives us a nice control for that. It's simple, doesn't require extra configuration, and doesn't require hunting down extra log files. The pytest docs actually encourage using printing this way.
https://docs.pytest.org/en/stable/capture.html#using-print-statements-for-debugging

That said, I get why people are wary of print statements. If there are strong opinions, I can switch it over.

File: tests/integration_tests/integration_settings.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (OddBloke) [reply tests/integration_tests/integration_settings.py:65]: > pycloudlib doesn't do anything with defaulting. You need to pass the values to the library itself through the various constructors.

We don't pass any parameters in `Ec2Client._get_client`, so _something_ below us must be handling defaulting (in that case, at least). My guess is that it's `boto` itself doing it; GCE's Python library does have similar support IIRC.

This seems like an inconsistency in `pycloudlib`'s interface that we need to iron out: either we should aim to not perform any defaulting, or we should use it wherever available; I think the mixture here is going to lead to confused developers.

> I wouldn't want to automatically use a pre-configured default though.

As mentioned, we already are doing this with EC2. Broadly speaking, I think this is expected behaviour from people, so I don't know if we need to actively avoid it. (Even more broadly speaking, people may even expect this behaviour _because_ it's how boto operates.)

> accidentally typed pytest in the root directory without understanding the implications

Yeah, this would be very easy to trip over even for those of us who do understand the implications and is orthogonal to defaulting: even if I've configured $expensive_cloud explicitly, I wouldn't want to launch instances just by running `pytest`. I wonder if there's anything we can do to exclude them from the "default" collect.

File: tests/integration_tests/conftest.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    if integration_settings.IMAGE_SOURCE == 'NONE':
+        pass  # that was easy
+    elif integration_settings.IMAGE_SOURCE == 'IN_PLACE':
+        if not isinstance(client, LxdContainerClient):
+            raise ValueError('IN_PLACE as IMAGE_SOURCE only works for LXD')
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/conftest.py:49]: Meeeeeh, six of one half dozen of the other to me. I get that instance checking is frowned upon in OO code, but sometimes it's easier to do a simple check for a one-off than to add attributes that only matter to a single type.

File: tests/integration_tests/integration_settings.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/integration_settings.py:65]: Yes, I see the inconsistency now. We can update pycloudlib to use configured defaults (though my personal preference is to make settings explicit). As far as the default collect, it looks like you can put pytest defaults in `tox.ini`. There is an exclude directory option. I'll play around with that.

File: tests/integration_tests/integration_settings.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+# Required for Oracle
+OCI_COMPARTMENT_ID = None
+
+##################################################################
+# USER SETTINGS OVERRIDES
+##################################################################
+# Bring in any user-file defined settings
+try:
+    from tests.integration_tests.user_settings import *
+except:  # noqa
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/integration_settings.py:81]: `except:` -> `except ImportError:` ?

File: tests/integration_tests/platforms.py [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+            '{sudo} apt-get update -q && '
+            '{sudo} apt-get install -qy cloud-init'
+        ).format(sudo='sudo' if self.use_sudo else '', repo=repo)
+        self._install_new_cloud_init(remote_script)
+
+    def install_deb(self):
+        print('Installing deb package')
+        deb_path = integration_settings.IMAGE_SOURCE
+        deb_name = os.path.basename(deb_path)
+        remote_path = '/tmp/{}'.format(deb_name)
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/platforms.py:139]: Can it run after we've SSH'ed? This could be a problem during an SRU, right?

File: tests/integration_tests/conftest.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (OddBloke) [reply tests/integration_tests/conftest.py:44]: > I have yet to see a use-case for multiple log levels in tests.

Agreed.

> I either want everything, or I want nothing.

When it comes to integration tests, I think we always want everything. Even if a test is going to pass, I want to see that things are happening while it gets to that point.

> The -s flag gives us a nice control for that. It's simple, doesn't require extra configuration, and doesn't require hunting down extra log files.

If we configure logging to log to stdout then I think we could probably continue to use `-s`, give or take?

> The pytest docs actually encourage using printing this way. https://docs.pytest.org/en/stable/capture.html#using-print-statements-for-debugging

The example given there feels different to what we're doing. Specifically, I think that "running this module will show you precisely the output of the failing function and hide the other one" is not desirable behaviour: as mentioned above, I think we really want to emit all this information all the time, regardless of pass/fail.

> That said, I get why people are wary of print statements. If there are strong opinions, I can switch it over.

In terms of what we output, I'm not particularly fussed about it being the logging library specifically, but I do think (a) timestamped output is a killer feature for long-running tests, and (b) it is very useful to know where in the code a particular message is coming from. (For (b): if I'm debugging tests then I can just `grep` for the message, but if I'm just observing them in CI, say, then having the extra context is really helpful.) Of course, `logging` is an included battery that does these things.

Another factor to consider: our dependencies do use `logging`, so even if we don't use it, we still need to be able to configure it for debugging lower-level issues etc. My experience has been that you don't want to only do this locally (i.e. by manually configuring logging in your local tree) because the sort of CI failures that need trace-level logging to understand generally only reproduce in the CI environment.

File: tests/integration_tests/conftest.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (OddBloke) [reply tests/integration_tests/conftest.py:49]: Yeah, we can leave this as-is for now: I can imagine that `multipass` or LXD VMs may also have support for this, as well as the fact that `LXDContainerClient` actually can't always support it, but let's cross that bridge when we come to it.

File: tests/integration_tests/integration_settings.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (OddBloke) [reply tests/integration_tests/integration_settings.py:81]: Aha, right, I had forgotten that it might not be present! `ImportError` would still mask if I had an incorrect `import` in my `user_settings.py`, but I think that's an acceptable corner case.

File: tests/integration_tests/integration_settings.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+#   LXD/VM only. Mount the source code as-is directly into
+#   the container to override the pre-existing cloud-init code
+# PROPOSED
+#   Install from the Ubuntu proposed repo
+# <ppa repo>, e.g., ppa:cloud-init-dev/proposed
+#   Install from a PPA. It MUST start with 'ppa:'
+# <file path>
+#   A path to a valid package to be uploaded and installed
+# <image id>
+#   A pre-existing (platform dependent) image id
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/integration_settings.py:54]: This is a good idea, but I'm not sure I understand the problem as you explain it. The image id text gets passed directly pycloudlib as the image id to use. If there's an image named 'tests' or 'cloudinit' on the platform being launched, then we'll use it. If there isn't, then we'll get an exception.

But I like the idea in a future-proofing sense. If we need to add other `IMAGE_SOURCE` types that can contain text with no obvious way to distinguish between an `image id` and `something else`, using a prefix for the source type makes sense.

File: tests/integration_tests/platforms.py [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+        self.pull_file(remote_path, tmp_file.name)
+        with tmp_file as f:
+            contents = f.read()
+        return contents
+
+    def push_to_file(self, remote_path, contents: str):
+        with NamedTemporaryFile('w', delete=False) as tmp_file:
+            tmp_file.write(contents)
+        self.push_file(tmp_file.name, remote_path)
+        os.unlink(tmp_file.name)
```
[the remark this answers is not in the document]
Comment (OddBloke) [reply tests/integration_tests/platforms.py:100]: > Could there be problems with pushing a still open file? We'd have to at least flush first, correct?

Hmm, great point; probably and probably.

In that case, perhaps:

```py
with contextlib.ExitStack() as stack:
    stack.callback(os.unlink, tmp_file.name)
    self.push_file(...)
```

would do the trick?

(Also, thinking about it, with `delete=False`, using `NamedTemporaryFile` as a context manager doesn't do anything, so probably just `tmp_file = NamedTemporaryFile('w', delete=False)` would make it clearer that we aren't using the built-in cleanup.)

> This implementation is a little lazy, but I wasn't too concerned with occasionally leaving a file in /tmp/ in the case where the push fails.

The failure case I'm concerned about is on a Jenkins slave, if we don't notice failures for hours or days then we could end up with an awful lot of these files in /tmp (and if they're cloud-init debs then they're ~0.5MB, so they can add up pretty fast). Not a huge deal, agreed.

File: tests/integration_tests/platforms.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (OddBloke) [reply tests/integration_tests/platforms.py:139]: > Can it run after we've SSH'ed?

I believe it could (though I don't have a reproducing system to hand)

> This could be a problem during an SRU, right?

Potentially, but I think a `cloud-init status --wait` before doing anything else means that we're out of the danger window.

File: tests/integration_tests/integration_settings.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (OddBloke) [reply tests/integration_tests/integration_settings.py:54]: > This is a good idea, but I'm not sure I understand the problem as you explain it.

If I have a file named `new` in my local tree and an image named `new` in my cloud, it's ambiguous what I'm asking for with `IMAGE_SOURCE = "new"`: am I specifying a "<file path>" or an "<image id>"?

However, it looks like this comment may be out-of-date: I can't see any way in which `IMAGE_SOURCE` is actually used as an image ID; it looks like `OS_IMAGE` is used for that. Given that, I agree there isn't a problem with the code as written. (Perhaps `OS_IMAGE` was added to address the issue I'm describing, without removing that part of the comment?)

> But I like the idea in a future-proofing sense.

/me does his best stopped clock impression.

File: integration-requirements.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply integration-requirements.txt:8]: I created a card to capture this

File: tests/integration_tests/conftest.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+    """Skip tests on unsupported clouds.
+
+    A test can take any number of marks to specify the platforms it can
+    run on. If a platform(s) is specified and we're not running on that
+    platform, then skip the test. If platform specific marks are not
+    specified, then we assume the test can be run anywhere.
+    """
+    all_platforms = client_name_to_class.keys()
+    supported_platforms = set(all_platforms).intersection(
+        mark.name for mark in item.iter_markers())
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/conftest.py:24]: I created a card to further explore this

File: tests/integration_tests/integration_settings.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/integration_settings.py:67]: I created a card to capture this

File: tests/integration_tests/platforms.py [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+    def _get_client(self):
+        raise NotImplementedError
+
+    def _get_image(self):
+        if current_image:
+            return current_image
+        image_id = self.settings.OS_IMAGE
+        try:
+            image_id = self.client.released_image(self.settings.OS_IMAGE)
+        except (ValueError, IndexError):
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/platforms.py:53]: I created a card to capture this

File: tests/integration_tests/test_simple.py [hunk 19]
```diff
@@ -0,0 +1,50 @@
+# This is an example test file. It won't get merged
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/test_simple.py:1]: Created a card for this

File: tests/integration_tests/platforms.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/platforms.py:100]: How about this?
```python
    def write_to_file(self, remote_path, contents: str):
        with NamedTemporaryFile('w', delete=False) as tmp_file:
            tmp_file.write(contents)

        try:
            self.push_file(tmp_file.name, remote_path)
        finally:
            os.unlink(tmp_file.name)
```

I still want the `NamedTemporaryFile` context manager so the file gets closed automatically (otherwise I need a try...finally around the write). Your `ExitStack()` code seems like a convenient way to get a generic context manager, but it feels a bit golf-y to me and it wasn't immediately clear to me what was happening, so I think I prefer a try...finally.

I can't combine them into a single `ExitStack()` block because I need the file object to add a callback, but since it's a stack, the unlink would happen before the close.

File: HACKING.rst [hunk 20]
```diff
@@ -173,9 +173,18 @@ Cloud Config Modules
...
 
-Unit Testing
+.. _unit_testing:
+
+Testing
 ------------
 
+cloud-init has both unit tests and integration tests. Unit tests can
+be found in-tree alongside the source code, as well as
+at ``tests/unittests``. Integration tests can be at
```
Comment (OddBloke) [inline HACKING.rst:183]:
```suggestion
at ``tests/unittests``. Integration tests can be found at
```

File: doc/rtd/topics/cloud_tests.rst [hunk 21]
```diff
@@ -1,6 +1,6 @@
-*******************
-Integration Testing
-*******************
+************************
+Cloud tests (Deprecated)
+************************
```
Comment (OddBloke) [inline doc/rtd/topics/cloud_tests.rst:4]: Is it worth adding a pointer to `integration_tests.rst`?

File: doc/rtd/topics/integration_tests.rst [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+* Launch an instance on the specified test platform
+* Install the version of cloud-init under test
+* Run ``cloud-init clean`` on the instance so subsequent boots
+  resemble out of the box behavior
+* Take a snapshot of the instance to be used as a new image from
+  which new instances can be launched
+
+Instance Setup
+==============
+Test setup occurs between image setup and test execution. Test setup
```
Comment (OddBloke) [inline doc/rtd/topics/integration_tests.rst:32]: `s/Test/Instance/g`?

File: tests/integration_tests/platforms.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (OddBloke) [reply tests/integration_tests/platforms.py:100]: LMAO I had forgotten about the existence of try/finally apparently. Yeah, that's way better.

File: tests/integration_tests/integration_settings.py [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+#   code.
+# PROPOSED
+#   Install from the Ubuntu proposed repo
+# <ppa repo>, e.g., ppa:cloud-init-dev/proposed
+#   Install from a PPA. It MUST start with 'ppa:'
+# <file path>
+#   A path to a valid package to be uploaded and installed
+# <image id>
+#   A pre-existing (platform dependent) image id
+IMAGE_SOURCE = 'NONE'
```
Comment (OddBloke) [inline tests/integration_tests/integration_settings.py:57]: On further reflection, is `IMAGE_SOURCE` the right name here? It seems more like `CLOUD_INIT_SOURCE`.

**Review by OddBloke — COMMENTED** [review]
Quick doc review (plus another couple of thoughts elsewhere), thanks for that!

File: tests/integration_tests/platforms.py [hunk 24]
```diff
@@ -72,12 +72,12 @@ def launch(self):
...
         launch_args.update(self.launch_kwargs)
         self.instance = self.client.launch(**launch_args)
-        print('Launched instance: {}'.format(self.instance))
+        log.info('Launched instance: %s', self.instance)
 
     def destroy(self):
         self.instance.delete()
 
-    def exec(self, command):
+    def execute(self, command):
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/platforms.py:80]: @smoser It appears we can do both. These both work:
```
print(client.execute('cloud-init -v'))
print(client.execute(['/usr/bin/cloud-init', '-v']))
```

**Comment by TheRealFalcon** [comment]
@OddBloke I believe I have addressed all of your comments. Let me know if there is anything I missed or anything else you find. I also got rid of the example test file.

File: tox.ini [hunk 25]
```diff
@@ -139,8 +139,15 @@ deps =
 
 [pytest]
 # TODO: s/--strict/--strict-markers/ once xenial support is dropped
+testpaths = cloudinit tests/unittests
```
Comment (OddBloke) [inline tox.ini:142]: I wonder if this allows us to drop our `posargs` defaulting from the other tox environments.

**Review by OddBloke — APPROVED** [review]
Thanks for all the work on this James, it's looking really good now!

I think we have one outstanding question from @smoser around shell escaping etc., but I don't think that needs to block landing this code so we can start writing tests; any refactor it leads to will be fairly mechanical (and it's not like we have a huge number of tests).

**Comment by TheRealFalcon** [comment]
@OddBloke I responded to @smoser 's question, and pycloudlib already gives us this behavior. You can pass in either a string or a list, so I don't think escaping is a problem. Correct me if I'm missing something.

File: tests/integration_tests/platforms.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (smoser) [reply tests/integration_tests/platforms.py:79]: Are you sure?

Try this:

    >>> subprocess.check_output(['echo', 'hi mom!'])
    b'hi mom!\n'
    >>> subprocess.check_output(['echo', 'hi mom!'], shell=True)
    b'\n'

I suspect your client.execute(['echo', 'hi mom!']) will do something like the second.

If it got that right, try this:

    client.execute(['echo', """Please put "2 spaces" after a period.  You'll make more $$ when you grow up. Sincerely, your 8th grade teacher"""])

Its worth getting 'exec' right once.

[the remark this answers is not in the document]
Comment (smoser) [reply tests/integration_tests/platforms.py:79]: actually... it might do the right thing.
[pycloudlib.execute](https://github.com/canonical/pycloudlib/blob/master/pycloudlib/instance.py#L132) uses [shell_pack](
https://github.com/canonical/pycloudlib/blob/master/pycloudlib/util.py#L80).  Hooray!

Thanks @powersj

[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply tests/integration_tests/platforms.py:79]: Confirmed. 
```
class TestExample:
    def test_example(self, client):
        print(client.execute(['echo', 'hi mom!']))
        print(client.execute(['echo', """Please put "2 spaces" after a period.  You'll make more $$ when you grow up. Sincerely, your 8th grade teacher"""]))
```

will output
```
hi mom!
Please put "2 spaces" after a period.  You'll make more $$ when you grow up. Sincerely, your 8th grade teacher
```

[the remark this answers is not in the document]
Comment (powersj) [reply tests/integration_tests/platforms.py:79]: @smoser I had a really good teacher :)
